### ISSUE TITLE
[feature request] --watch with generate.js is broken after refactor

### ISSUE THREAD

**Issue by syzer** [typicode/json-server#160]
seriously? :)
will you accept pull request?

**Comment by typicode** [comment]
Sorry that you have problems with this feature.
It works for me on Linux at least:

```javascript
module.exports = function () {
  return {
    foo: []
  }
}
```

```
json-server generate.js --watch

  \{^_^}/ hi!

  Loading generate.js
  Done

  Resources
  http://localhost:3000/foo

  Home
  http://localhost:3000

  Type s + enter at any time to create a snapshot of the database
  Watching...

  generate.js has changed, reloading...

  Loading generate.js
  Done

  Resources
  http://localhost:3000/bar

  Home
  http://localhost:3000
```

I've changed `generate.js` to:

```javascript
module.exports = function () {
  return {
    bar: []
  }
}
```

And waited for something like 2-3 seconds.

What's your OS?

**Comment by syzer** [comment]
mac..

**Comment by typicode** [comment]
Ok thanks, will test on it :)

**Comment by syzer** [comment]
`../../node_modules/.bin/json-server -w --routes=routes.json --port=3002 generate.js`

**Comment by typicode** [comment]
Works too.

![screen shot 2015-07-30 at 5 16 56 pm](https://cloud.githubusercontent.com/assets/5502029/8986969/f125cdce-36de-11e5-9656-5395e9603683.png)

I'm on `v0.7.23`, that's strange.

**Comment by syzer** [comment]
i'm puzzled.. i'll jump on that

**Comment by syzer** [comment]
ok i figured out
we are not watching the directory (as it was before).. but file

please checkout my fork 
`git@github.com:syzer/json-server.git`
and run
`./bin/index.js --watch Trash/gen.js`

**Comment by typicode** [comment]
I understand better now. Thank you for the example :)
But right now, I prefer to keep watching a file instead of a directory.

Mainly for the sake of simplicity and to avoid side effects.
For example, considering this project structure:

```
/generate.js
/node_modules
/public/index.js
```

`generate.js` will be reloaded when a module is installed or when `index.js` is modified.
And I fear that in the end, `nodemon` will just be recoded in JSON Server.

So, instead as a workaround I would suggest this:

```javascript
// server.js
var jsonServer = require('json-server')
var generate = require('./generate')

var router = jsonServer.router(generate())

var server = jsonServer.create()
server.use(jsonServer.defaults)
server.use(router)

server.listen(3000)
```

```
nodemon server
```

Or using `node-touch`:

```javascript
// generate.js
fs.watch('./other', function () { touch(__filename) })
```

**Comment by syzer** [comment]
i made a fix for that to avoid workouraund, will you merge?

**Comment by typicode** [comment]
You mean this one https://github.com/typicode/json-server/commit/34434329fe2b54e7583338a6c8f5c4230ad6063c?

**Comment by syzer** [comment]
Almost, ill push it tomorrow morning.

On Fri, Jul 31, 2015 at 8:53 PM, typicode <notifications@github.com> wrote:

> You mean this one 3443432
> <https://github.com/typicode/json-server/commit/34434329fe2b54e7583338a6c8f5c4230ad6063c>
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/typicode/json-server/issues/160#issuecomment-126783161>
> .
>



-- 
--Lukas 'SyZer' Gintowt

**Comment by syzer** [comment]
^^^ ?

**Comment by typicode** [comment]
Thanks for the PR, but I think it's a complex problem.
And I'm sorry not being able to pull.

## Scenario 1

For example, I've cloned your repo and changed:

```javascript
  // src/cli/watch.js
  fs.watch(watchedDir, {recursive: true}, function (event, changedFile) {
    // Atom text editor is generating a 'change' event
    // So with just 'rename' cb was not called
    // if (event === 'rename') cb()
    cb()
  })
```

Then I've created:

```javascript
// db/generate.js
var other = require('./other')

module.exports = function () {
  return other()
}
```

```javascript
// db/other.js
module.exports = function () {
  return { foo: [] }
}
```

And started JSON Server using:

```
node bin --watch db/generate.js
```

If `other.js` is changed:

```javascript
// db/other.js
module.exports = function () {
  return { bar: [] }
}
```

`generate.js` will be reloaded but not `other.js` because it's cached by `require()` and there won't be a `http://localhost:3000/bar` route.

## Scenario 2

Another complicated scenario is if you have the following project structure:

```
generate.js
other.js
public/index.html
public/index.js
```

`public/index.js` will be watched too. So when `public/index.js` is changed it will trigger a reload of `generate.js` even though it's not required by `generate.js`.

**Comment by typicode** [comment]
Scenario 1

I've cloned https://github.com/syzer/json-server but couldn't make it reload `db/other.js` on Linux (maybe I'm missing something).

Scenario 2

Yes :) this is the main reason why I'm not in favor of watching a directory. I agree that it would be better to reload the JS file when a required file is changed.

However on the other hand and from a user point of view, I think it would be more "surprising" to see the server being reloaded because a `public/index.js` is changed.

In other words, the fix would introduce another problem.

**Comment by syzer** [comment]
Scenario 1
maybe that is because i clear require cache in generate.js
do not know either :)

Scenario 2 
you are right.
i will probably use the gnu `watch`.
